src/methods/validate-fixture-input.ts

```diff
@@ -124,14 +172,44 @@ export function validateFixtureInput(
             }
           }
 
+          const namedType = getNamedType(fieldType);
+          let possibleTypes: string[] = [];
+          if (isLeafType(namedType)) {
+            // do nothing
+          } else if (isAbstractType(namedType)) {
+            possibleTypes = schema.getPossibleTypes(namedType).map(type => type.name);
+          } else if (isObjectType(namedType)) {
+            possibleTypes = [namedType.name];
+          }
+
+          possibleTypesStack.push(new Set(possibleTypes));
+          typeStack.push(getNamedType(fieldType));
+
           valueStack.push(nestedValues);
         },
         leave() {
           valueStack.pop();
+          typeStack.pop();
+          possibleTypesStack.pop();
         },
       },
       SelectionSet: {
         enter(node) {
+          // Look ahead to find __typename field and track its response key
+          const typenameField = node.selections.find(
+            (selection) =>
+              selection.kind === Kind.FIELD &&
+              selection.name.value === "__typename"
+          );
+
+          // If this SelectionSet has __typename, use its response key.
+          // Otherwise, inherit from parent.
+          const typenameResponseKey = typenameField && typenameField.kind === Kind.FIELD
+            ? typenameField.alias?.value || "__typename"
+            : typenameResponseKeyStack[typenameResponseKeyStack.length - 1];
```

Comment:
Inheriting from the parent is only correct if the selection is part of an inline fragment, not part of a field. Is there a way we can differentiate? Maybe split this out to entering inline fragments and fields? Or I think all of the arguments to `enter` are actually `node, key, parent, path, ancestors`, so we could include those in the signature to get access to the parent to check what it is.